src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -255,7 +277,7 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
             hasErrors = hasErrors || iterationVariableType.IsErrorType();
 
             // Skip the conversion checks and array/enumerator differentiation if we know we have an error (except local name conflicts).
-            if (hasErrors)
+            if (hasErrors || _syntax.HasErrors)
```

Comment:
I'll change the syntax errors check to check for syntax errors only on the declaration part.